### ISSUE TITLE
Esbuild fail to bundle from 4.10 to 4.12

### ISSUE THREAD

**Issue by richard-julien** [antlr/antlr4#4211]
- [x] I have reproduced my issue using the latest version of ANTLR
- [x] I have done a search of the existing issues to make sure I'm not sending in a duplicate

Please include the following information:
 - target information: **Javascript**
 - smallest possible grammar and code that reproduces the behavior: **N/A related only to code import**
 - description of the expected behavior and actual behavior: No error bundling code with esbuild

Hi Antlr team,

I will try to expose my problem in a clearer way but sorry if information looks like partial. Difficult to explain correctly my problem with my current knowledge of ANTLR.

I try to upgrade ANTLR in my project (nodejs) from 4.10 to 4.12.
My project use esbuild to bundle the source code.
After upgrading to 4.12, the bundled code refuse to start with the error.

```
node:internal/modules/cjs/loader:1367
    throw new ERR_INVALID_ARG_VALUE('filename', filename, createRequireError);
    ^

TypeError [ERR_INVALID_ARG_VALUE]: The argument 'filename' must be a file URL object, file URL string, or absolute path string. Received undefined
```

Looking into the bundled code, its looks like the esbuild transform the file and i have this line.

`var ha = (0, import_module.createRequire)(import_meta.url)("fs");`

And as the error said, the import_meta is empty initialized so url is undefined.

Looking into esbuild, the documentation said that import.meta.url is not supported when bundling:

`You also may not want to bundle your dependencies with esbuild. There are many node-specific features that esbuild doesn't support while bundling such as __dirname, import.meta.url, fs.readFileSync, and *.node native binary modules. You can exclude all of your dependencies from the bundle by setting [packages](https://esbuild.github.io/api/#packages) to external:`

So my question is, do you change something in ANTLR that now use the meta import feature that break the compatibility with esbuild bundling? 

Thanks in advance.

**Comment by ericvergnaud** [comment]
Yes the packaging changed in 4.12 for TypeScript to work.Envoyé de mon iPhoneLe 28 mars 2023 à 11:18, Julien Richard ***@***.***> a écrit :﻿

 I have reproduced my issue using the latest version of ANTLR
 I have done a search of the existing issues to make sure I'm not sending in a duplicate

Please include the following information:

target information: Javascript
smallest possible grammar and code that reproduces the behavior: N/A related only to code import
description of the expected behavior and actual behavior: No error bundling code with esbuild

Hi Antlr team,
I will try to expose my problem in a clearer way but sorry if information looks like partial. Difficult to explain correctly my problem with my current knowledge of ANTLR.
I try to upgrade ANTLR in my project (nodejs) from 4.10 to 4.12.
My project use esbuild to bundle the source code.
After upgrading to 4.12, the bundled code refuse to start with the error.
node:internal/modules/cjs/loader:1367
    throw new ERR_INVALID_ARG_VALUE('filename', filename, createRequireError);
    ^

TypeError [ERR_INVALID_ARG_VALUE]: The argument 'filename' must be a file URL object, file URL string, or absolute path string. Received undefined

Looking into the bundled code, its looks like the esbuild transform the file and i have this line.
var ha = (0, import_module.createRequire)(import_meta.url)("fs");
And as the error said, the import_meta is empty initialized so url is undefined.
Looking into esbuild, the documentation said that import.meta.url is not supported when bundling:
You also may not want to bundle your dependencies with esbuild. There are many node-specific features that esbuild doesn't support while bundling such as __dirname, import.meta.url, fs.readFileSync, and *.node native binary modules. You can exclude all of your dependencies from the bundle by setting [packages](https://esbuild.github.io/api/#packages) to external:
So my question is, do you change something in ANTLR that now use the meta import feature that break the compatibility with esbuild bundling?
Thanks in advance.

**Comment by richard-julien** [comment]
Any solution/workaround available @ericvergnaud? Problem is still present in 4.13 and i dont know where to start to upgrade.

**Comment by Tarjei400** [comment]
@richard-julien How you build your app? If you could provide some small repo with your setup that reproduces this error I can assist you with that.

**Comment by ericvergnaud** [comment]
This might already be fixed, a release is planned this week-end, let's see if it happens and does fix the problem.

**Comment by Tarjei400** [comment]
@ericvergnaud Cool! Just in case let me know guys if there are issues, would gladly help :+1:

**Comment by ShaharCider** [comment]
@ericvergnaud Hey, has this been released yet?

**Comment by ericvergnaud** [comment]
Can you try manually replacing antlr4's package.json with latest from 'development' branch and see of it fixes the problem ?

**Comment by ericvergnaud** [comment]
The dev file seems to work for many people, I have never used esbuild so I encourage you to try out things
Maybe what's missing is a esbuild cli option ?
also not sure what target.js is (filename error raises all sorts of suspicions)

**Comment by ericvergnaud** [comment]
The generated parser requires the antlr runtime, regardless of packaging

**Comment by eirueirufu** [comment]
@ZenMagus i met the same problem, it seems to be caused by the import.meta.url, You can see if this solution can help you https://github.com/evanw/esbuild/issues/1492#issuecomment-893144483

**Comment by ericvergnaud** [comment]
All, please try 4.13.1-patch-1, so we can possibly close this.

**Comment by Janther** [comment]
The bug is still in this repo but I found a way to inject a fix at the command level in this comment :

https://github.com/evanw/esbuild/issues/2441#issuecomment-1315341698

**Comment by DavyLandman** [comment]
I've managed to work around it by patching the `package.json` by replacing all the `antlr.*.mjs` with `antlr.*.cjs` in the exports section.

so it became:

```
 "node": {
        "import": "./dist/antlr4.node.cjs",
        "require": "./dist/antlr4.node.cjs",
        "default": "./dist/antlr4.node.cjs"
      },
```

as `esbuild` kept picking up the `mjs` files, even though it was running in commonjs mode.
